rts/lra/lra-test/lra-test-arquillian-extension/src/main/java/io/narayana/lra/arquillian/spi/NarayanaLRARecovery.java

```diff
@@ -0,0 +1,67 @@
+package io.narayana.lra.arquillian.spi;
+
+import io.narayana.lra.LRAConstants;
+import io.narayana.lra.client.NarayanaLRAClient;
+import io.narayana.lra.logging.LRALogger;
+import org.eclipse.microprofile.lra.tck.service.spi.LraRecoveryService;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.client.WebTarget;
+import javax.ws.rs.core.Response;
+import java.net.URI;
+
+@ApplicationScoped
+public class NarayanaLRARecovery implements LraRecoveryService {
+
+    @Override
+    public void triggerRecovery() {
+        doTriggerRecovery(null, System.getProperty(NarayanaLRAClient.LRA_COORDINATOR_HOST_KEY),
+            Integer.parseInt(System.getProperty(NarayanaLRAClient.LRA_COORDINATOR_PORT_KEY)));
+    }
+
+    @Override
+    public void triggerRecovery(URI lraId) {
+        doTriggerRecovery(lraId, lraId.getHost(), lraId.getPort());
+    }
+
+    private void doTriggerRecovery(URI lraId, String host, int port) {
+        if (invokeCoordinatorRecovery(lraId, host, port)) {
+            // first recovery scan probably collided with periodic recovevery which started
+            // before the test execution so try once more
+            invokeCoordinatorRecovery(null, host, port);
+        }
+    }
+
+    /**
+     * Invokes LRA coordinator recovery REST endpoint and returns whether the recovery of intended LRA happended
+     *
+     * @param lraId LRA id of the LRA that is expected to recover
+     * @param host the LRA coordinator host address
+     * @param port the LRA coordinator port
+     * @return true if the expected LRA recovered and no repeated recovery is necessary, false otherwise
+     */
+    private boolean invokeCoordinatorRecovery(URI lraId, String host, int port) {
+        // trigger a recovery scan
+        Client recoveryCoordinatorClient = ClientBuilder.newClient();
+
+        try {
+            String recoveryCoordinatorUrl = String.format("http://%s:%d/%s/recovery",
+                host, port, LRAConstants.RECOVERY_COORDINATOR_PATH_NAME);
+            WebTarget recoveryTarget = recoveryCoordinatorClient.target(URI.create(recoveryCoordinatorUrl));
+
+            // send the request to the recovery coordinator
+            Response response = recoveryTarget.request().get();
+            String json = response.readEntity(String.class);
+            LRALogger.logger.error(lraId);
+            LRALogger.logger.error(json);
+            response.close();
+
+            return lraId != null && json.contains(lraId.toASCIIString());
```

Comment:
The code still looks obscure and a reader would still like to know:

1. If you call recovery without an LRA what is the meaning of returning false from this method.
2. If you call recovery without an LRA why do you need to call recovery twice.

Reply:
ok, I will pass it to both invocations and just ignore the return value in the second call.

Reply:
The javadoc says "true if the expected LRA recovered" which does not reflect what the code does. This makes the code in `invokeRecovery(URI lraId, String host, int port)` look wrong.

I think it's best to avoid the overloading of the calls to `invokeRecoveryCoordinatorEndpoint` since that is what is making the code appear confusing. Just call the `invokeRecoveryCoordinatorEndpoint` method from both of the interface methods - that way it will be clear to the reader what the intention is.

Also the interface methods should be responsible for deciding whether or not to trigger recovery. The `triggerRecovery(URI lraId)` can return a boolean to indicate whether or not lraId was recovered and the `triggerRecovery()` should either return a void or a count of recovering LRAs or a list of them. But my preference would be to return a void.

Reply:
Or returning count has the benefit of letting the caller know whether there is anything left to recover which is more useful that void.